### PR TITLE
nixos/anbox: use mainline drivers when available

### PR DESCRIPTION
**Pull request by mvnetbiz**
###### Motivation for this change
virtualisation.anbox is broken on Linux 5.x

###### Things done
- updated Anbox package to latest version
- marked anbox kernel modules broken for Linux >= 5.7. They do not build because Linux 5.7 unexported kallsyms_lookup_name()
 - enable CONFIG_ options for the mainline kernel drivers needed
 - change Anbox NixOS module to use the correct drivers if using Linux 5.x

I have confirmed this working with Linux 4.19, 5.4, 5.7

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/os-specific/linux/anbox/default.nix
Comment (mvnetbiz) [inline pkgs/os-specific/linux/anbox/default.nix:87]: I am not sure about the build inputs. I do not think they were separated correctly before, but I think some of them should still be moved to nativeBuildInputs. Are the snapcraft build-packages/stage-packages dependencies listed here https://github.com/anbox/anbox/blob/7f1bfaf90de1f808d4ce8f49ef050120c5c8d1ef/snap/snapcraft.yaml#L262-L310 analogous to nativeBuildInputs/buildInputs? Is there a way to sandbox dependencies of a Nix package at run time to test this or do you just have to know? I see at least pkgconfig I am sure is a nativeBuildInput.

Comment (SuperSandro2000) [reply pkgs/os-specific/linux/anbox/default.nix:87]: According to their reference for snapcarft.yaml yes but I am not sure if this translates 1:1 to NixOS.

https://snapcraft.io/docs/snapcraft-yaml-reference

Comment (FRidh) [reply pkgs/os-specific/linux/anbox/default.nix:87]: Please set `strictDeps = true;`, that should help identifying which are `nativeBuildInputs`.

**Comment by mvnetbiz** [comment]
I've fixed the build inputs, fixed up some stuff like .desktop file and app menu category, and added a test for the service module.

**Comment by mvnetbiz** [comment]
Rebased and fixed the nixos test to be reliable. Are there any concerns since this changes the kernel?

File: nixos/modules/virtualisation/anbox.nix
Comment (spinus) [inline nixos/modules/virtualisation/anbox.nix:139]: don't you have to add `Type=fork` if you use daemon here?
You are adding it only to avoid the warning or any other reason?

**Comment by spinus** [comment]
@mvnetbiz thanks, that looks pretty good.

File: nixos/modules/virtualisation/anbox.nix
Comment (mvnetbiz) [reply nixos/modules/virtualisation/anbox.nix:139]: I did add it to avoid the warning but the process doesn't fork anyways, or at least doesn't exit.

**Comment by mvnetbiz** [comment]
I don't know what considerations should be taken for adding drivers to the default kernel, or if there is an alternative option.

**Review by Atemu — APPROVED** [review]
Amazing work!

Working as ~~well~~poorly as upstream allows AFAICT. Tested on linux_zen 5.9. Networking works too.

**Comment by bqv** [comment]
Ping @mvnetbiz

File: nixos/modules/virtualisation/anbox.nix
Comment (mvnetbiz) [reply nixos/modules/virtualisation/anbox.nix:79]: As far as I can tell these rules don't do anything anyways.

Comment (mvnetbiz) [reply nixos/modules/virtualisation/anbox.nix:79]: I can't recall but they either don't run or anbox works correctly with the default mode anyways.

**Comment by mvnetbiz** [comment]
Oops, accidentally re based on wrong branch. @bqv hello.

**Comment by bqv** [comment]
@SuperSandro2000 pong

File: nixos/tests/anbox.nix
Comment (mvnetbiz) [reply nixos/tests/anbox.nix:20]: Will that just cause an evaluation error or any other problems? I am assuming it is still okay to keep the test here.

**Comment by Mic92** [comment]
@bqv @Atemu Has someone of you tested the lastest rebase?

File: pkgs/os-specific/linux/kernel/common-config.nix
Comment (Mic92) [inline pkgs/os-specific/linux/kernel/common-config.nix:461]: Is it possible to build them as module?

Comment (spinus) [reply pkgs/os-specific/linux/kernel/common-config.nix:461]: does it make sense to add nixos test here?

Comment (Atemu) [reply pkgs/os-specific/linux/kernel/common-config.nix:461]: Not anymore I think. That broke with 5.7.

Although, I'm not sure whether the modules were available in 5.0 yet.

Comment (Atemu) [reply pkgs/os-specific/linux/kernel/common-config.nix:461]: @spinus this PR already adds a NixOS test

Comment (mvnetbiz) [reply pkgs/os-specific/linux/kernel/common-config.nix:461]: @Mic92 Zen kernel has patches that enable building as module.
https://github.com/zen-kernel/zen-kernel/commit/a628884e152edd7dd987eb989ddc107a2e8af713
https://github.com/zen-kernel/zen-kernel/commit/eda591e5a88289287362cadefda66f68c580800f

**Comment by bqv** [comment]
> @bqv @Atemu Has someone of you tested the lastest rebase?

Yes. beem using it since bqv/nixrc@0257af13. Fine for me

**Comment by mvnetbiz** [comment]
I am looking into building the drivers as modules using the patches from linux-zen. Should I add it to a reasonable few kernels in `pkgs/top-level/all-packages.nix` or have it conditionally include the patch based on the NixOS module being enabled?

**Comment by bqv** [comment]
If it's a module, I see no harm in adding it unconditionally. Caching is a nice perk from that

File: pkgs/os-specific/linux/kernel/common-config.nix
Comment (Mic92) [reply pkgs/os-specific/linux/kernel/common-config.nix:461]: I asked the zen-kernel maintainers about upstreaming this. I think we should adopt the patch for now - Many people won't need those modules by default.

**Comment by Mic92** [comment]
> I am looking into building the drivers as modules using the patches from linux-zen. Should I add it to a reasonable few kernels in `pkgs/top-level/all-packages.nix` or have it conditionally include the patch based on the NixOS module being enabled?

I think we could include them in newer kernels but maybe also look into bringing them upstream.

**Comment by willfaust** [comment]
I was just trying to install anbox kernel modules and was wondering why it failed. I came here to see that there is in fact a pr, but it still hasn't been merged yet... any reason why? I would really like to have a working anbox. No rush, but hopefully this comes to the notice of people and it gets merged asap. Thanks!

**Review by IvarWithoutBones — APPROVED** [review]

**Comment by mvnetbiz** [comment]
I will take a look at it again. I think we wanted to build the drivers as modules, but I couldn't get a patch to apply to all kernel versions or something.

**Comment by Mic92** [comment]
> Could we merge this so anbox is functional, while waiting for a PR that uses module patches?

Yes. please leave out the kernel changes for now.

**Comment by mvnetbiz** [comment]
Okay, I'll work on the kernel patches separately. I'm unable to build 5.10 or 5.11 with the module patches at this point anyways. zen kernel has also reverted the patches. https://github.com/zen-kernel/zen-kernel/commit/feb930b92396859a09ad5db7d0dbaba0fe34e496

**Comment by mvnetbiz** [comment]
> > Could we merge this so anbox is functional, while waiting for a PR that uses module patches?
> 
> Yes. please leave out the kernel changes for now.

Do you mean to merge this without any kernel changes at all?  Or to just build in the drivers with =y, and worry about making them build as modules later. I think you mean the latter since anbox won't work at all without the drivers?

**Comment by mvnetbiz** [comment]
Testing changes before pushing now. It would be super cool to use robotnix instead of the prebuild images for this at some future point.

**Comment by mvnetbiz** [comment]
NixOS test driver with linuxPackages_5_11 seems to hang at
`machine # [    6.357398] systemd[970]: Started D-Bus User Message Bus.`
but testing manually, it works.

**Comment by Mic92** [comment]
> > > Could we merge this so anbox is functional, while waiting for a PR that uses module patches?
> > 
> > 
> > Yes. please leave out the kernel changes for now.
> 
> Do you mean to merge this without any kernel changes at all? Or to just build in the drivers with =y, and worry about making them build as modules later. I think you mean the latter since anbox won't work at all without the drivers?

Would it be reasonable to require the zen kernel that for anbox user's and hope that eventually we get module support for android related drivers?

**Comment by Atemu** [comment]
I think the zen kernel doesn't ship the patches anymore.

You shouldn't need any special patches though, we could just build anbox/android support into our regular kernels. When I tested this back in december, anbox simply worked after mounting the binderfs IIRC.  
The zen patches were for building the android stuff as a module instead of baking it in AFAICT.

**Comment by mvnetbiz** [comment]
> I think the zen kernel doesn't ship the patches anymore.
> 
> You shouldn't need any special patches though, we could just build anbox/android support into our regular kernels. When I tested this back in december, anbox simply worked after mounting the binderfs IIRC.
> The zen patches were for building the android stuff as a module instead of baking it in AFAICT.

Yes, this is all correct. I think the preference was to have it built as a module to theoretically reduce attack surface on kernel of users who don't use anbox. It looks like there have previously been a couple of CVEs for these subsystems. Maybe it is better to leave this open for now, but it works as-is if someone wants to pick it themselves.

**Comment by Mic92** [comment]
I think it is also possible to set kernel options in the anbox module itself.

**Comment by Atemu** [comment]
Zen kernel with these patches worked just now. Currently compiling vanilla 5.11...

> theoretically reduce attack surface on kernel of users who don't use anbox.

I think protecting themself against theoretical attack vectors should be left to the user. Someone who cares about things like that would have build their own kernels anyways.

> I think it is also possible to set kernel options in the anbox module itself by setting boot.kernelPatches

It would but that'd mean local kernel builds after activating anbox which isn't a great UX.  
We'd also have to add a channel blocking hydra job for this and I'm not sure how comfortable I'd feel blocking on something as unstable as e.g. an anbox test (which couldn't run in hydra anyways in its current form because of adb).

**Review by Atemu — APPROVED** [review]
Tested on linux_lqx and linux_latest.

**Comment by bqv** [comment]
diverged, it seems

**Comment by Atemu** [comment]
@mvnetbiz do you want to continue this PR and rebase it?

I'd also highly recommend giving [marvin](https://github.com/timokau/marvin-mk2/) a try.

**Comment by mvnetbiz** [comment]
Rebased.
Fixed anbox package, there is a bug with lxc 4.0.7+.

**Comment by egasimus** [comment]
How to enable this on NixOS `unstable`?

**Comment by bqv** [comment]
Wait for it to be merged, or patch it into your nixpkgs

**Comment by bqv** [comment]
conflict

**Comment by mvnetbiz** [comment]
errr what

File: nixos/modules/virtualisation/anbox.nix
Comment (jonringer) [inline nixos/modules/virtualisation/anbox.nix:9]:
```suggestion
  useAnboxModules = kernelPackages.kernelOlder "5.0";
```

File: nixos/tests/anbox.nix
Comment (samueldr) [reply nixos/tests/anbox.nix:20]: (Will probably adapt to the new PR)

I think it's time we have a non-SDK package for adb, fastboot and friends.

And there is one!

 - https://github.com/nmeum/android-tools

This is currently in use by Alpine Linux

 - https://git.alpinelinux.org/aports/tree/community/android-tools/APKBUILD

Comment (Atemu) [reply nixos/tests/anbox.nix]: https://github.com/NixOS/nixpkgs/pull/124992

**Comment by mvnetbiz** [comment]
It looks like it didn't notifiy automatically, but I think I will close this in favor of #125600, which includes more fixes and some new options for image customization as well, and hopefully the NixOS test from here can be integrated, which now includes the standalone adb.

**Comment by Lillecarl** [comment]
Can't wait to see this land, good job! :)

**Review by VergeDX — APPROVED** [review]

**Pull request closed without merging**
